### DISCUSSION TITLE
OAuth support for model serving on Databricks

### DISCUSSION THREAD

**Discussion by ishaan-mehta** [mlflow/mlflow#11300] [Q&A] [answered]
I noticed that @victorsun123 created a PR that was merged yesterday (https://github.com/mlflow/mlflow/pull/11129) that enables MLflow to refresh an OAuth token when getting credentials for model serving on Databricks.

It looks like this requires the OAuth token being stored in `/var/credentials-secret/model-dependencies-oauth-token` — how does one store an OAuth token there in the first place?

Is there an environment variable containing the OAuth secret or token that should be passed to the model serving endpoint when deploying a model to the endpoint? I can't find any other mention of this file path in the repository.

Thank you for your assistance!

**Comment by victorsun123** [comment]
Hey Ishaan, this is for an upcoming improvement where we plan to simplify credential workflows for requests out of model serving environment.  In this change, we will automatically parse metadata and check ACLs for dependencies when model is logged, then when model is served, we will auto populate and refresh the OAuth token in that location that can be read using the  `databricks_utils` function here. Nothing to be done on the users end! 

Right now, support for this is going to be limited to authorizing Databricks vector search, but plan to add support for this for other model dependencies. Curious to learn more about what your use case is?  Thanks!

**Reply by ishaan-mehta** [reply]
Hey @victorsun123 — thanks for the info!

My current use case, where I am running into an issue, is deploying a RAG model (utilizing Databricks vector search) to a model serving endpoint *using only service principal credentials*.

MLflow's latest release only has support for authenticating with a host and token (it cannot authenticate with a client ID and client secret) due to its dependency on the legacy Databricks CLI (which only supports PAT-based authentication). So, when I deploy my model to the model serving endpoint, I can only pass a generated OAuth token (using the service principal's OAuth secret) as `DATABRICKS_TOKEN`, which is fine only until the token expires in one hour. This is what I mean:

```
latest_model_version = get_latest_model_version(MODEL_NAME)

w = WorkspaceClient()

endpoint_config = EndpointCoreConfigInput(
    name=MODEL_NAME,
    served_models=[
        ServedModelInput(
            model_name=MODEL_NAME,
            model_version=get_latest_model_version(MODEL_NAME),
            workload_size="Small",
            scale_to_zero_enabled=True,
            environment_vars={
                "DATABRICKS_TOKEN": "{{secrets/spn/oauth_token}}"
            },
        )
    ],
)
```

where the `oauth_token` secret contains an OAuth token recently generated via either REST API or `curl`. After the token expires, the endpoint is no longer functional.

If you do the same thing but with the `DATABRICKS_CLIENT_SECRET` environment variable, querying the endpoint results in a `You haven't configured the CLI yet` error because MLflow fails to set up authentication, as it uses the legacy Databricks CLI to authenticate, which does not support OAuth authentication.

If you replace the secret with a user PAT, the endpoint will work as long as the PAT is valid, but we are trying to avoid using user credentials.

So, I was wondering if this PR you made would resolve this issue where you are required to either use a 1-hour OAuth token or use a user token when deploying to a model serving endpoint.

Sorry for the long reply, just wanted to provide the context fully. 🙂

**Reply by victorsun123** [reply]
Ahh yes, this is targeting your exact use case! 

Assuming you are setting up your RAG model via `mlflow.langchain.log_model`  you can use any authentication method for initializing/playing with the the vector search client in dev. Behind the scenes, we will parse all the metadata and on the model serving side, we will constantly refresh the OAuth token for that vector search object at the path referenced in the mlflow PR. At query time, the vector search client will call this method to get the OAuth credentials, no setup or configuration needed on the user's side, and no need to worry about token expiration.

NOTE: this feature is still in development and there are still some changes that need to be made on the service/backend side before this end-to-end workflow will work as intended, so will update you and provide more detailed documentation when this feature is launched! 

cc @aakrati @dbczumar

**Reply by ishaan-mehta** [reply]
Oh great, thanks, appreciate the quick response & info!

Just to make sure we are talking about the same thing, though, since you mentioned this is only going to support authorizing vector search:

The error I mentioned occurs regardless of whether we are using vector search or not. It seems the issue is caused by a fundamental limitation of how MLflow is able to authenticate with a Databricks workspace (it can only use PAT-based authentication due to its dependency on the legacy Databricks CLI). In other words, it seems unrelated to the vector search client's authentication — MLflow _itself_ is unable to authenticate in the model serving environment unless you pass a token in `DATABRICKS_TOKEN` when deploying the model to the model serving endpoint.

See this [Databricks notebook source file (compressed b/c GitHub doesn't allow uploading .py files)](https://github.com/mlflow/mlflow/files/14466206/Authentication.MRE.v7.1.py.zip) for how to reproduce the error when deploying a simple LangChain chain (without vector search) to a model serving endpoint. When you query the endpoint created by this notebook, it will say that `You haven't configured the CLI yet`.
<img width="881" alt="image" src="https://github.com/mlflow/mlflow/assets/45380942/d4e796e6-f8b0-444a-9f55-feddd9048e01">

As I mentioned, currently you can only use a user PAT or a generated OAuth token (that lasts an hour) to get around this error (passing it as a secret environment variable as shown in my comment above).

So will this feature allow MLflow to authenticate with Databricks in the model serving environment by passing an OAuth secret instead of passing `DATABRICKS_TOKEN` (user PAT or temporary OAuth token) when deploying the model?

**Reply by victorsun123** [reply, marked as answer]
That's correct, by default MLFLow has no way of knowing credentials needed to access your specific resources whether that be vector database, other models, or other resources. (Exception is feature tables which supports automatic credentials lookup only when used with the feature store client) Right now we are working on adding parsing of metadata to check permissions and generate credentials for Langchain MLFlow models using vector search, and hope to add support for connecting to other models soon as well. 

The workflow today for most customers is similar to how you have setup, that you have to manually configure credentials via environment variables, which is not ideal, as OAuth token will expire in an hour, and PAT token has a variety of security concerns, given unlimited scope and TTL. This feature will automatically handle the credentials so user does not have to worry about configuring environment variables, token expiration, etc.

**Reply by ishaan-mehta** [reply]
Awesome — thanks for the information and clarification, Victor! Looking forward to this. 🙂

**Reply by ishaan-mehta** [reply]
So is it fair to say that until this feature goes live, there isn’t any way to pass a service principal OAuth secret or service principal client secret to a model serving endpoint to authenticate usage of a Databricks foundation model (e.g., Mixtral)? In other words, the only way to deploy a LangChain model that uses Mixtral is to either use a PAT or a temporary OAuth token?

If so, is there currently any guidance from Databricks on how to handle service-principal-based model deployment for LangChain models using Mixtral until the new feature goes live? Perhaps one of the following options?

- Pass a user PAT as an environment variable when deploying a model to an endpoint.
- Use a Databricks-managed service principal so that a PAT can be generated on its behalf and passed as an environment variable when deploying a model to an endpoint. (It seems AAD/Entra ID-managed service principals cannot have PATs generated on their behalf.)

**Reply by victorsun123** [reply]
Yeah unfortunately I think generating a PAT and configuring as an environment variable might be your best solution right now. Will update you when we're able to improve this workflow!
